proposals/POEM-2-function-pulling-container-scheduler.md

```diff
@@ -0,0 +1,132 @@
+* [Architecture](https://cwiki.apache.org/confluence/display/OPENWHISK/System+Architecture)
+* [Component Design](https://cwiki.apache.org/confluence/display/OPENWHISK/Component+Design)
+
+### Implementation details
+
+For the record, we(Naver) are already operating OpenWhisk with this scheduler in a production environment.
+We want to contribute the scheduler and hope it evolves with the community.
+
+There are many [new components](https://cwiki.apache.org/confluence/display/OPENWHISK/Component+Design).
+
+#### Common components
+We store data in ETCD, there are many relevant components such as `EtcdClient`, `LeaseKeepAliveService`, `WatcherService`, `DataManagementService`, etc.
+
+* `EtcdClient`: An ETCD client offering basic CRUD operations.
+* `DataManagementService`: In charge of storing/recovering data in ETCD. It internally utilizes `LeaseKeepAliveService`, and `WatcherService`.
+* `LeaseKeepAliveService`: In charge of keeping the given lease alive.
+* `WatcherService`: Watches the given keys and receives events for the keys.
+
+
+#### Scheduler components
+
+In an abstract view, schedulers provide queueing, container scheduling, and activation routing.
+
+* `QueueManager`: The main entry point for queue creation request. It has references to all queues.
+* `MemoryQueue`: Dynamically created/deleted for each action. It watches the incoming/outgoing requests and triggers container creation.
+* `ContainerManager`: Schedule container creation requests to appropriate invokers.
+* `ActivationServiceImpl`: Provide API for containers to fetch activations via Akka-grpc. It works in a long-poll way to avoid busy-waiting.
+
+#### Controller components
+* `FPCPoolBalancer`: Create queues if not exist, and forward messages to them.
+* `FPCEntitlementProvider`: Throttle activations based on throttling information in ETCD.
+
+#### Invoker components.
+
+* `FunctionPullingContainerPool`: A container pool for function pulling container. It handles the container creation requests.
+* `FunctionPullingContainerProxy`: A proxy for a container. It repeatedly fetches activations and invokes them.
+* `ActivationClientProxy`: The Akka-grpc client. It communicates with `ActivationServiceImpl` in schedulers.
+* `InvokerHealthManager`: Manages the health and resources information of invokers. The data is stored in ETCD. If an invoker becomes unhealthy, it invokes health activations.
+
+## Future work.
+
+#### Persistence
+
+We reached a conclusion that the persistence of activation requests is not that mandatory requirement along with the at-most-once nature and circuit-breaking of OpenWhisk.
+But if it is desired, we can implement a persistent queue rather than an in-memory queue.
+
+#### Multiple partitions
+
+Currently, a queue only has one partition. Since there can be multiple queues for each action, cluster-wide RPS(request per second) linearly increases.
```

Comment:
```suggestion
Currently, a queue only has one partition. Since there can be multiple queues for each action, cluster-wide RPS (request per second) increases linearly.
```